File: src/ie/states/go.ts
import { removeNonDigits } from "../../shared/normalize.js";
import { weightedSum, mod11CheckDigit } from "../../shared/mod11.js";

// Verified against the official SEFAZ-GO "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_GO.html), including its worked
// example (10.987.654-7). Format: AB.CDE.FGH-I, where AB must be 10, 11, or
// 20-29.
const WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2];

function normalize(value: string): string {
  return removeNonDigits(value);
}

function hasValidPrefix(ie: string): boolean {
  const prefix = Number(ie.slice(0, 2));

  return prefix === 10 || prefix === 11 || (prefix >= 20 && prefix <= 29);
}

function isValid(value: string): boolean {
  if (!/^[\d.\-\s]+$/.test(value)) {
    return false;
  }

  const ie = normalize(value);

  if (ie.length !== 9 || !hasValidPrefix(ie)) {
    return false;
  }

  return mod11CheckDigit(weightedSum(ie, WEIGHTS)) === Number(ie[8]);
}

function format(value: string): string {
  const ie = normalize(value);

  if (ie.length !== 9) {
    return value;
  }

  return ie.replace(/^(\d{2})(\d{3})(\d{3})(\d)$/, "$1.$2.$3-$4");
}

export const GO = {
  isValid,
  normalize,
  format,
};

File: csharp/src/BrValidator/Ie/Go.cs
using System.Text.RegularExpressions;
using BrValidator.Internal;

namespace BrValidator.Ie
{
    /// <summary>
    /// GO: verified against the official SEFAZ-GO "Roteiro de Crítica da
    /// Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_GO.html),
    /// including its worked example (10.987.654-7). Format: AB.CDE.FGH-I,
    /// where AB must be 10, 11, or 20-29.
    /// </summary>
    internal static class Go
    {
        private static readonly Regex AllowedChars = new Regex(@"^[\d.\-\s]+$", RegexOptions.Compiled);
        private static readonly int[] Weights = { 9, 8, 7, 6, 5, 4, 3, 2 };

        public static string Normalize(string value) => Shared.RemoveNonDigits(value);

        private static bool HasValidPrefix(string v)
        {
            int prefix = int.Parse(v.Substring(0, 2));

            return prefix == 10 || prefix == 11 || (prefix >= 20 && prefix <= 29);
        }

        public static bool IsValid(string value)
        {
            if (!AllowedChars.IsMatch(value))
            {
                return false;
            }

            string v = Normalize(value);

            if (v.Length != 9 || !HasValidPrefix(v))
            {
                return false;
            }

            return Mod11.Mod11CheckDigit(Mod11.WeightedSum(v, Weights)) == (v[8] - '0');
        }

        public static string Format(string value)
        {
            string v = Normalize(value);

            if (v.Length != 9)
            {
                return value;
            }

            return $"{v.Substring(0, 2)}.{v.Substring(2, 3)}.{v.Substring(5, 3)}-{v.Substring(8, 1)}";
        }
    }
}
